File: library/src/main/java/com/github/easyguide/layer/GuideLayerView.java
package com.github.easyguide.layer;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.RelativeLayout;

import com.github.easyguide.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by shenxl on 2018/8/16.
 */

public class GuideLayerView extends RelativeLayout {
    private static final int DEFAULT_BASE_COLOR = 0x60000000;
    private int mBaseColor = DEFAULT_BASE_COLOR;
    private List<Rect> mTargetRects = new ArrayList<>();
    private List<View> mExtraViews = new ArrayList<>();
    private Paint mPaint;
    private PorterDuffXfermode mXfermode;
    private boolean mHasMarginReset;
    private DrawCallBack mDrawCallBack;
    private LayerClickListener mLayerClickListener;
    private float mDownX, mDownY;

    public GuideLayerView(Context context) {
        super(context);
        init();
    }

    public GuideLayerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public GuideLayerView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public GuideLayerView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    private void init() {
        mPaint = new Paint();
        mXfermode = new PorterDuffXfermode(PorterDuff.Mode.CLEAR);
        setLayerType(LAYER_TYPE_SOFTWARE, null);
    }

    public void setBaseColor(int baseColor) {
        mBaseColor = baseColor;
    }

    void setDrawCallBack(DrawCallBack drawCallBack) {
        mDrawCallBack = drawCallBack;
    }

    public void setLayerClickListener(LayerClickListener layerClickListener) {
        mLayerClickListener = layerClickListener;
    }

    void addTargetsRect(Rect rect) {
        mTargetRects.add(rect);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(MeasureSpec.makeMeasureSpec(MeasureSpec.getSize(widthMeasureSpec), MeasureSpec.EXACTLY),
                MeasureSpec.makeMeasureSpec(MeasureSpec.getSize(heightMeasureSpec), MeasureSpec.EXACTLY));
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int childCount = this.getChildCount();
        for (int i = 0; i < childCount; i++) {
            View child = this.getChildAt(i);
            Location location = (Location) child.getTag();
            if (location.index >= mTargetRects.size()){
                continue;
            }
            Rect targetRect = mTargetRects.get(location.index);
            int verticalAxis = (targetRect.left + targetRect.right) / 2;
            int horizontalAxis = (targetRect.top + targetRect.bottom) / 2;
            int width = child.getMeasuredWidth();
            int height = child.getMeasuredHeight();
            switch (location) {
                case TO_TOP:
                    child.layout(verticalAxis - width / 2, targetRect.top - height, verticalAxis + width / 2, targetRect.top);
                    break;
                case TO_BOTTOM:
                    child.layout(verticalAxis - width / 2, targetRect.bottom, verticalAxis + width / 2, targetRect.bottom + height);
                    break;
                case TO_LEFT:
                    child.layout(targetRect.left - width, horizontalAxis - height / 2, targetRect.left, horizontalAxis + height / 2);
                    break;
                case TO_RIGHT:
                    child.layout(targetRect.right, horizontalAxis - height / 2, targetRect.right + width, horizontalAxis + height / 2);
                    break;
                default:
                    child.layout(targetRect.left, targetRect.top, targetRect.right, targetRect.bottom);
                    break;
            }
        }
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
        canvas.drawColor(mBaseColor);
        mPaint.reset();
        mPaint.setAntiAlias(true);
        mPaint.setXfermode(mXfermode);
        if (mDrawCallBack != null) {
            for (int i = 0; i < mTargetRects.size(); i++) {
                mDrawCallBack.onDraw(i, mTargetRects.get(i), canvas, mPaint);
            }
        }
        super.dispatchDraw(canvas);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int action = event.getAction();
        switch (action) {
            case MotionEvent.ACTION_DOWN:
                mDownX = event.getX();
                mDownY = event.getY();
                return true;
            case MotionEvent.ACTION_MOVE:
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                float upX = event.getX();
                float upY = event.getY();
                if (Math.abs(upX - mDownX) < 10 && Math.abs(upY - mDownY) < 10) {
                    if (mLayerClickListener != null) {
                        mLayerClickListener.onFullClick();
                        for (int i = 0; i < mTargetRects.size(); i++) {
                            if (contains(mTargetRects.get(i), upX, upY)) {
                                mLayerClickListener.onSingleClick(i);
                                return true;
                            }
                        }
                    }
                    performClick();
                }
                break;

        }
        return super.onTouchEvent(event);
    }

    private boolean contains(Rect rect, float x, float y) {
        return rect.left < rect.right && rect.top < rect.bottom  // check for empty first
                && x >= rect.left && x < rect.right && y >= rect.top && y < rect.bottom;
    }

    interface DrawCallBack {
        void onDraw(int id, Rect rect, Canvas canvas, Paint paint);
    }

    interface LayerClickListener {
        void onFullClick();

        void onSingleClick(int id);
    }
}

File: library/src/main/java/com/github/easyguide/layer/GuideLayerView.kt
package com.github.easyguide.layer

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.widget.RelativeLayout

/**
 * Created by shenxl on 2018/8/16.
 */

class GuideLayerView : RelativeLayout {
    var baseColor = 0x60000000
    private val targetRects = mutableListOf<Rect>()
    private val paint = Paint()
    private val xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
    internal var drawCallBack: ((index: Int, rect: Rect, canvas: Canvas, paint: Paint) -> Unit)? = null
    internal var targetClickListener: ((index: Int) -> Unit)? = null
    private var downX = 0f
    private var downY = 0f

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    init {
        setLayerType(View.LAYER_TYPE_SOFTWARE, null)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(View.MeasureSpec.makeMeasureSpec(View.MeasureSpec.getSize(widthMeasureSpec), View.MeasureSpec.EXACTLY),
                View.MeasureSpec.makeMeasureSpec(View.MeasureSpec.getSize(heightMeasureSpec), View.MeasureSpec.EXACTLY))
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        val childCount = this.childCount
        for (i in 0 until childCount) {
            val child = this.getChildAt(i)
            val locBean = child.tag as LocBean
            if (locBean.targetIndex !in 0 until targetRects.size) {
                continue
            }
            val targetRect = targetRects[locBean.targetIndex]
            val verticalAxis = (targetRect.left + targetRect.right) / 2
            val horizontalAxis = (targetRect.top + targetRect.bottom) / 2
            val width = child.measuredWidth
            val height = child.measuredHeight
            var top = 0
            var bottom = 0
            var left = 0
            var right = 0
            locBean.locs.forEach {
                when (it) {
                    Location.TO_TOP -> {
                        left = if (left == 0) verticalAxis - width / 2 else left
                        top = targetRect.top - height
                        right = if (right == 0) verticalAxis + width / 2 else right
                        bottom = targetRect.top
                    }
                    Location.TO_BOTTOM -> {
                        left = if (left == 0) verticalAxis - width / 2 else left
                        top = targetRect.bottom
                        right = if (right == 0) verticalAxis + width / 2 else right
                        bottom = targetRect.bottom + height
                    }
                    Location.TO_LEFT -> {
                        left = targetRect.left - width
                        top = if (top == 0) horizontalAxis - height / 2 else top
                        right = targetRect.left
                        bottom = if (bottom == 0) horizontalAxis + height / 2 else bottom
                    }
                    Location.TO_RIGHT -> {
                        left = targetRect.right
                        top = if (top == 0) horizontalAxis - height / 2 else top
                        right = targetRect.right + width
                        bottom = if (bottom == 0) horizontalAxis + height / 2 else bottom
                    }
                    Location.COVER -> {
                        left = targetRect.left
                        top = targetRect.top
                        right = targetRect.right
                        bottom = targetRect.bottom
                    }
                    Location.ALIGN_TOP -> {
                        left = if (left == 0) verticalAxis - width / 2 else left
                        top = targetRect.top
                        right = if (right == 0) verticalAxis + width / 2 else right
                        bottom = targetRect.top + height
                    }
                    Location.ALIGN_BOTTOM -> {
                        left = if (left == 0) verticalAxis - width / 2 else left
                        top = targetRect.bottom - height
                        right = if (right == 0) verticalAxis + width / 2 else right
                        bottom = targetRect.bottom
                    }
                    Location.ALIGN_LEFT -> {
                        left = targetRect.left
                        top = if (top == 0) horizontalAxis - height / 2 else top
                        right = targetRect.left + width
                        bottom = if (bottom == 0) horizontalAxis + height / 2 else bottom
                    }
                    Location.ALIGN_RIGHT -> {
                        left = targetRect.right - width
                        top = if (top == 0) horizontalAxis - height / 2 else top
                        right = targetRect.right
                        bottom = if (bottom == 0) horizontalAxis + height / 2 else bottom
                    }
                }
            }
            child.layout(left + locBean.horizontalOffset,
                    top + locBean.verticalOffset,
                    right + locBean.horizontalOffset,
                    bottom + locBean.verticalOffset)
        }
    }

    override fun dispatchDraw(canvas: Canvas) {
        canvas.drawColor(baseColor)
        paint.reset()
        paint.isAntiAlias = true
        paint.xfermode = xfermode
        drawCallBack?.let {
            for ((index, value) in targetRects.withIndex()) {
                it.invoke(index, value, canvas, paint)
            }
        }
        super.dispatchDraw(canvas)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val action = event.action
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                downX = event.x
                downY = event.y
                return true
            }
            MotionEvent.ACTION_UP -> {
                performClick()
                val upX = event.x
                val upY = event.y
                if (Math.abs(upX - downX) < 10 && Math.abs(upY - downY) < 10) {
                    targetClickListener?.let {
                        for ((index, value) in targetRects.withIndex()) {
                            if (value.contains(upX, upY)) {
                                it.invoke(index)
                                return true
                            }
                        }
                        it.invoke(-1)
                        return true
                    }
                }
            }
        }
        return super.onTouchEvent(event)
    }

    internal fun addTargetsRect(rect: Rect) {
        targetRects.add(rect)
    }

    internal fun addExtraView(view: View, targetIndex: Int, verticalOffset: Int, horizontalOffset: Int, locs: List<Location>) {
        view.tag = LocBean(targetIndex, locs, verticalOffset, horizontalOffset)
        addView(view)
    }

    private fun Rect.contains(x: Float, y: Float): Boolean {
        return (left < right && top < bottom && x >= left && x < right && y >= top && y < bottom)
    }

    private data class LocBean(var targetIndex: Int, var locs: List<Location>, var verticalOffset: Int, var horizontalOffset: Int)
}
